src/week2.hs

```diff
@@ -0,0 +1,176 @@
+{-# OPTIONS_GHC -Wall #-}
+--module week2 where
+
+-- Mastermind -----------------------------------------
+
+-- A peg can be one of six colors
+data Peg = Red | Green | Blue | Yellow | Orange | Purple
+         deriving (Show, Eq, Ord)
+
+-- A code is defined to simply be a list of Pegs
+type Code = [Peg]
+
+-- A move is constructed using a Code and two integers; the number of
+-- exact matches and the number of regular matches
+data Move = Move Code Int Int
+          deriving (Show, Eq)
+
+-- List containing all of the different Pegs
+colors :: [Peg]
+colors = [Red, Green, Blue, Yellow, Orange, Purple]
+
+-- Exercise 1 -----------------------------------------
+
+-- Get the number of exact matches between the actual code and the guess
+exactMatches :: Code -> Code -> Int
+
+exactMatches [] _ = 0
+exactMatches _ [] = 0
+exactMatches (x:xs) (x':xs') 
+    |x == x' = 1 + (exactMatches xs xs')
+    |otherwise = (exactMatches xs xs')
+
+-- Exercise 2 -----------------------------------------
+-- Write a function that returns the number of total matches between the secret code and the guess
+
+-- helper
+sumLists :: [Int] -> [Int] -> [Int]
+sumLists [] _ = []
+sumLists _ [] = []
+sumLists (x:xs) (x':xs') = (x + x'):(sumLists xs xs')
+
+--or, with implicit recursion (Jan's suggestion):
+sumLists1 :: [Int] -> [Int] -> [Int]
+sumLists1 a b = zipWith (+) a b
+
+countColors :: Code -> [Int]
+
+countColors [] = [0,0,0,0,0,0]
+countColors (x:xs)
+    | x == Red    = sumLists [1,0,0,0,0,0] (countColors xs)
+    | x == Green  = sumLists [0,1,0,0,0,0] (countColors xs)
+    | x == Blue   = sumLists [0,0,1,0,0,0] (countColors xs)
+    | x == Yellow = sumLists [0,0,0,1,0,0] (countColors xs)
+    | x == Orange = sumLists [0,0,0,0,1,0] (countColors xs)
```

Comment:
when you're case splitting across constructors (rather than some more ad hoc kind of conditions), the most natural thing is to probably use a case expression (scroll to "case expressions" [here](http://learnyouahaskell.com/syntax-in-functions))

Reply:
another thing you can try for fun---and this would probably be overkill for this specific situation, but if say you had say 64 colors, you'd strongly prefer to do it---is to derive an `Enum` instance ([documentation](https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Enum.html#t:Enum)) for the `Peg` class and then use it to define a function
```haskell
basis :: Peg -> [Int]
```
that returns the correct `[0,..,1,...,0]` list, solely referencing the `Enum` instance, so as to avoid hard-coding them. This is definitely not necessary for this problem but I think it could be a fun "challenge" :)

Reply:
Thanks for the comments! Yes, using an encode function like this is a lot better. I've added a new version of countColors